src/main/java/graphql/nadel/engine/AsyncIsFieldForbidden.java

```diff
@@ -32,9 +32,32 @@ public AsyncIsFieldForbidden(ServiceExecutionHooks serviceExecutionHooks, NadelC
     }
 
     public CompletableFuture<Map<String, GraphQLError>> getForbiddenFields(Node<?> root) {
-        return visitChildren(root).thenApply(unused -> {
-            return fieldIdsToErrors;
-        });
+        return CompletableFuture.allOf(
+                checkTopLevelField(root),
+                visitChildren(root)
+        ).thenApply(unused -> fieldIdsToErrors);
+    }
+
+    private CompletableFuture<Void> checkTopLevelField(Node root) {
```

Comment:
I think you should compose this code instead of copy pasting the code. This function looks very similar to `visitField`.

e.g.

```
CompleteableFuture<void> visitField(Node<?> node) {
    if (!(node instanceof Field)) {
        return CompletableFuture.completedFuture(null);
    }
    Field field = (Field) node;

    …

    return serviceExecutionHooks.isFieldForbidden(field, normalizedFields, fieldDefinitionOverall, nadelContext.getUserSuppliedContext())
        .whenComplete((graphQLError, throwable) -> {
            if (graphQLError != null && graphQLError.isPresent()) {
                fieldIdsToErrors.put(fieldId, graphQLError.get());
            }
        });
}

CompletableFuture<Void> visitFieldAndChildren(Node<?> node) {
        return visitField(node).thenCompose((error) -> {
            if (error.isPresent()) {
                return CompletableFuture.completedFuture(null);
            }
            return visitChildren(node);
        });
}
```

I think it's possible, but let me know if you think otherwise.

Reply:
applied